File: practices/interview-bit/stacks-and-queues/redundant-braces.py
class Solution:
    # @param A: string
    # @return an integer
    def braces(self, A):
        s = []
        ops = set(['+','-','*','/'])
        for c in A:
            if c == '(':
                s.append([c, 0])
            elif c == ')':
                if not s:
                    return 1  # Invalid braces
                if s[-1][1] == 0:
                    return 1
                s.pop()
            else:
                if s and (c in ops):
                    s[-1][1] = 1
        return 0

File: practices/interview-bit/stacks-and-queues/redundant-braces.cpp
// O(N) in time
// O(N) in space
bool isOperator(char c) {
    return c == '+' || c == '*' || c == '-' || c == '/';
}

int Solution::braces(string A) {
    // `s` will only contain '(' character
    // ( '(', 0 ) means there is no expression/operator in the current pair of braces
    stack<pair<char, int>> s;
    for(char c : A) {
        if (c == '(') {
            s.push(make_pair(c, 0));
        } else if (c == ')') {
            if (s.empty()) return 1;  // Invalid braces
            if (s.top().second == 0) return 1;  // Inside this pair of braces,
                                                // there is no expression, thus
                                                // redundant
            s.pop();
        } else {
            if (isOperator(c) && !s.empty()) {
                s.top().second = 1;
            }
        }
    }
    if (!s.empty()) return 1;
    return 0;
}
